### PR TITLE
restore ISerializer ability, "safely"

### LINKED ISSUE #185: v5 - Microsoft.Web.Redis.ISerializer is not available any more
As for v5 Microsoft.Web.Redis.ISerializer is not available any more, now how to implement custom Serializer?

### PR DESCRIPTION
**Pull request by zachrybaker**
This will resolve [issue 185](https://github.com/Azure/aspnet-redis-providers/issues/185)

### CODE REVIEW

**Review by stanleysmall-microsoft — COMMENTED** [review]
Thanks for this contribution! I think people will like being able to use custom serialization

**Comment by stanleysmall-microsoft** [comment]
@zachrybaker can you please agree to the CLA?

File: src/OutputCacheProvider/RedisOutputCacheProvider.csproj [hunk 1]
```diff
@@ -25,13 +25,13 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNet.OutputCache.OutputCacheModuleAsync" Version="1.0.2" />
-    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="5.0.0" />
+    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="8.0.0" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
-    <PackageReference Include="StackExchange.Redis" Version="2.6.122" />
-    <PackageReference Include="System.Diagnostics.PerformanceCounter" Version="6.0.1" />
-    <PackageReference Include="System.IO.Pipelines" Version="6.0.2" />
+    <PackageReference Include="StackExchange.Redis" Version="2.7.4" />
```
Comment (philon-msft) [inline src/OutputCacheProvider/RedisOutputCacheProvider.csproj:30]: A newer v2.7.10 of SE.Redis is now available

File: src/RedisSessionStateProvider/KeyGenerator.cs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 //
 
 using System;
+using System.Web.SessionState;
```
Comment (philon-msft) [inline src/RedisSessionStateProvider/KeyGenerator.cs:7]: Is this new `using` statement necessary?

File: src/Shared/BinaryFormattingSessionSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+﻿using System.IO;
+using System.Web.SessionState;
+
+namespace Microsoft.Web.RedisSessionStateProvider
+{
+    /// <summary>
+    /// The default, "safe" serialization.  
+    /// Hint. It's not safe. But it is what MS allows today.
```
Comment (philon-msft) [inline src/Shared/BinaryFormattingSessionSerializer.cs:8]: Comments should be professional

File: src/Shared/ISessionDataSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Web.SessionState;
+
+namespace Microsoft.Web.RedisSessionStateProvider
+{
+    public interface ISessionDataSerializer
```
Comment (philon-msft) [inline src/Shared/ISessionDataSerializer.cs:5]: Interface members should all have doc comments

File: src/Shared/BinaryFormattingSessionSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    /// Better is to use json or MessagePack.
+    /// </summary>
+    internal class BinaryFormattingSessionSerializer : ISessionDataSerializer
+    {
+        public string StorageTypeName { get => "SessionStateItemCollection"; }
```
Comment (philon-msft) [inline src/Shared/BinaryFormattingSessionSerializer.cs:13]:
```suggestion
        public string StorageTypeName => "SessionStateItemCollection";
```

File: src/RedisSessionStateProvider/KeyGenerator.cs [hunk 6]
```diff
@@ -13,17 +14,19 @@ internal class KeyGenerator
         public string DataKey { get; private set; }
         public string LockKey { get; private set; }
         public string InternalKey { get; private set; }
+        public string SessionDataType { get; private set; }
 
         private void GenerateKeys(string id, string app)
         {
             this.id = id;
-            DataKey = $"{{{app}_{id}}}_SessionStateItemCollection";
+            DataKey = $"{{{app}_{id}}}_{SessionDataType}";
```
Comment (philon-msft) [inline src/RedisSessionStateProvider/KeyGenerator.cs:22]: We should retain the "SessionStateItemCollection" portion of the key unless we need to eliminate it. What's your intention for the new `SessionDataType`? I assume to tag keys with a serializer identifier?

File: src/OutputCacheProvider/RedisOutputCacheProvider.csproj (hunk 1 above)
Comment (stanleysmall-microsoft) [reply src/OutputCacheProvider/RedisOutputCacheProvider.csproj]: all packages have been updated

File: src/RedisSessionStateProvider/KeyGenerator.cs (hunk 2 above)
Comment (stanleysmall-microsoft) [reply src/RedisSessionStateProvider/KeyGenerator.cs]: removed

File: src/Shared/BinaryFormattingSessionSerializer.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+            BinaryReader reader = new BinaryReader(ms);
+            return SessionStateItemCollection.Deserialize(reader);
+
+        }
+
+        public byte[] Serialize(SessionStateItemCollection data)
+        {
+            MemoryStream ms = new MemoryStream();
+            BinaryWriter writer = new BinaryWriter(ms);
+            ((SessionStateItemCollection)data).Serialize(writer);
```
[the remark this answers is not in the document]
Comment (stanleysmall-microsoft) [reply src/Shared/BinaryFormattingSessionSerializer.cs]: fixed

File: src/Shared/BinaryFormattingSessionSerializer.cs (hunk 3 above)
Comment (stanleysmall-microsoft) [reply src/Shared/BinaryFormattingSessionSerializer.cs]: updated

File: src/Shared/BinaryFormattingSessionSerializer.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+using System.Web.SessionState;
+
+namespace Microsoft.Web.RedisSessionStateProvider
+{
+    /// <summary>
+    /// The default, "safe" serialization.  
+    /// Hint. It's not safe. But it is what MS allows today.
+    /// Better is to use json or MessagePack.
+    /// </summary>
+    internal class BinaryFormattingSessionSerializer : ISessionDataSerializer
```
[the remark this answers is not in the document]
Comment (stanleysmall-microsoft) [reply src/Shared/BinaryFormattingSessionSerializer.cs]: done

File: src/Shared/ISessionDataSerializer.cs (hunk 4 above)
Comment (stanleysmall-microsoft) [reply src/Shared/ISessionDataSerializer.cs]: added

File: src/Shared/BinaryFormattingSessionSerializer.cs (hunk 5 above)
Comment (stanleysmall-microsoft) [reply src/Shared/BinaryFormattingSessionSerializer.cs]: removed StorageTypeName altogether.

File: src/RedisSessionStateProvider/KeyGenerator.cs (hunk 6 above)
Comment (stanleysmall-microsoft) [reply src/RedisSessionStateProvider/KeyGenerator.cs]: This change has been reverted

**Review by stanleysmall-microsoft — DISMISSED** [review]

**Comment by zachrybaker** [comment]
> @zachrybaker please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
>  > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by zachrybaker** [comment]
@microsoft-github-policy-service agree

**Comment by zachrybaker** [comment]
@stanleysmall-microsoft if you can recap what you’re looking for me at this point, I can try to put some attention on it over the weekend or something.

**Comment by zachrybaker** [comment]
Yes, the intention is to note how "THIS" instance of the data in redis is
serialized, vs some other one. I have a totally different Blazor app also
syncing this data.
Zachry Baker

File: src/Shared/ProviderConfiguration.cs [hunk 9]
```diff
@@ -87,6 +90,8 @@ private ProviderConfiguration(NameValueCollection config)
             // All below parameters are only fetched from web.config
             DatabaseId = GetIntSettings(config, "databaseId", 0);
             ApplicationName = GetStringSettings(config, "applicationName", null);
+            PublishSessionChanges = GetBoolSettings(config, "publishSessionChanges", false);
```
[the remark this answers is not in the document]
Comment (stanleysmall-microsoft) [reply src/Shared/ProviderConfiguration.cs]: @zachrybaker can you briefly describe the motivation behind adding the `publishSessionChanges` parameter? I don't think we should add it in this PR, but I'm curious about what this does.

**Comment by stanleysmall-microsoft** [comment]
> @stanleysmall-microsoft if you can recap what you’re looking for me at this point, I can try to put some attention on it over the weekend or something.

@zachrybaker thanks for making this PR. I think your changes will improve the user experience of the package. I already made some changes to your branch which I think can be merged soon. The only questions I had were about `publishSessionChanges` and the renaming the session key. Maybe we could add these features in a separate PR? I'm just trying to understand the impact of these potential changes for now.

File: src/OutputCacheProvider/RedisOutputCacheProvider.csproj [hunk 10]
```diff
@@ -24,14 +24,14 @@
...
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNet.OutputCache.OutputCacheModuleAsync" Version="1.0.2" />
-    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="5.0.0" />
+    <PackageReference Include="Microsoft.AspNet.OutputCache.OutputCacheModuleAsync" Version="1.0.3" />
+    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="8.0.0" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
-    <PackageReference Include="StackExchange.Redis" Version="2.6.122" />
-    <PackageReference Include="System.Diagnostics.PerformanceCounter" Version="6.0.1" />
-    <PackageReference Include="System.IO.Pipelines" Version="6.0.2" />
+    <PackageReference Include="StackExchange.Redis" Version="2.7.10" />
```
Comment (philon-msft) [inline src/OutputCacheProvider/RedisOutputCacheProvider.csproj]: Update the SE.Redis references to the new 2.7.17 version?

Comment (stanleysmall-microsoft) [reply src/OutputCacheProvider/RedisOutputCacheProvider.csproj]: Done

**Comment by zachrybaker** [comment]
Yes.

I added the publish feature so that another app/process can subscribe to
the update.

My use case is that I am incrementally rewriting a large web forms app in
Blazor and for most of that time period they will share session and profile
data. Taking the strangler fig approach where both apps will coexist and
the Blazor version will proxy unhandled requests to the old app. Since
Blazor does not really have the HTTP request/response model but is a
circuit - more of a current state machine, I needed a mechanism to
synchronize profile and session between the two and this publish fits the
bill perfectly.

File: src/RedisSessionStateProvider/KeyGenerator.cs (hunk 6 above)
Comment (philon-msft) [reply src/RedisSessionStateProvider/KeyGenerator.cs]: I think we should still include something like a "SerializerId" token in key names _in addition to "SessionStateItemCollection"_

File: src/Shared/ProviderConfiguration.cs [hunk 11]
```diff
@@ -117,6 +119,22 @@ private ProviderConfiguration(NameValueCollection config)
...
             OperationTimeoutInMilliSec = GetIntSettings(config, "operationTimeoutInMilliseconds", 0);
+
+            var serializationTypeNameSpace = GetStringSettings(config, "sessionSerializationNamespaceAndType", null);
+            var serializationTypeAssembly = GetStringSettings(config, "sessionSerializationTypeAssembly", null);
+
+            if (!string.IsNullOrEmpty(serializationTypeNameSpace) && !string.IsNullOrEmpty(serializationTypeAssembly))
+            {
+                try
+                {
+                    var serializer = Activator.CreateInstance(serializationTypeAssembly, serializationTypeNameSpace);
```
Comment (philon-msft) [inline src/Shared/ProviderConfiguration.cs]: This is a very slow/dangerous/ugly way to instantiate things. Can callers pass their instance of `ISessionDataSerializer` to `ProviderConfiguration.SessionDataSerializer` instead of having us instantiate it?

File: src/Shared/DefaultSessionStateSerializer.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
+﻿using System.IO;
+using System.Web.SessionState;
+
+namespace Microsoft.Web.RedisSessionStateProvider
```
Comment (philon-msft) [inline src/Shared/DefaultSessionStateSerializer.cs]: Why are these files in the `Shared` directory? They're not used by the OutputCacheProvider, right?

File: test/OutputCacheProviderFunctionalTests/RedisOutputCacheProvider.FunctionalTests.csproj [hunk 13]
```diff
@@ -16,19 +16,19 @@
     <ProjectReference Include="..\..\src\OutputCacheProvider\RedisOutputCacheProvider.csproj" />
```
Comment (philon-msft) [inline test/OutputCacheProviderFunctionalTests/RedisOutputCacheProvider.FunctionalTests.csproj]: Changes to the OutputCacheProvider should be in a separate PR if possible

File: src/Shared/ProviderConfiguration.cs [hunk 14]
```diff
@@ -28,6 +29,7 @@ internal class ProviderConfiguration
         public int ConnectionTimeoutInMilliSec { get; set; }
         public int OperationTimeoutInMilliSec { get; set; }
         public string ConnectionString { get; set; }
+        public ISessionDataSerializer SessionDataSerializer { get; set; } = new DefaultSessionStateSerializer();
```
Comment (philon-msft) [inline src/Shared/ProviderConfiguration.cs]: This SessionDataSerializer property is not relevant to the ProviderConfiguration, right? If so, it shouldn't be added to this `ProviderConfiguration` shared by Session and Output cache providers. Instead, we should created dedicated `ProviderConfiguration`s for each (using a shared base class)

File: src/Shared/StackExchangeClientConnection.cs [hunk 15]
```diff
@@ -188,9 +188,9 @@ internal SessionStateItemCollection DeserializeSessionStateItemCollection(RedisR
         {
             try
             {
-                MemoryStream ms = new MemoryStream((byte[])serializedSessionStateItemCollection);
-                BinaryReader reader = new BinaryReader(ms);
-                return SessionStateItemCollection.Deserialize(reader);
+                var bytes = (byte[])serializedSessionStateItemCollection;
```
Comment (philon-msft) [inline src/Shared/StackExchangeClientConnection.cs]: These "SessionState"-specific methods also don't belong in this "Shared" file. They should be moved into a file under `/RedisSessionStateProvider`

File: test/RedisSessionStateProviderFunctionalTests/RedisSessionStateProvider.FunctionalTests.csproj [hunk 16]
```diff
@@ -16,19 +16,19 @@
     <ProjectReference Include="..\..\src\RedisSessionStateProvider\RedisSessionStateProvider.csproj" />
```
Comment (philon-msft) [inline test/RedisSessionStateProviderFunctionalTests/RedisSessionStateProvider.FunctionalTests.csproj:16]: We should add a functional test that provides its own `ISessionDataSerializer` to test that scenario, and demonstrate how others should do it

File: src/Shared/DefaultSessionStateSerializer.cs (hunk 12 above)
Comment (stanleysmall-microsoft) [reply src/Shared/DefaultSessionStateSerializer.cs]: These files have been moved

File: src/Shared/ProviderConfiguration.cs (hunk 14 above)
Comment (stanleysmall-microsoft) [reply src/Shared/ProviderConfiguration.cs]: Done

File: src/Shared/StackExchangeClientConnection.cs (hunk 15 above)
Comment (stanleysmall-microsoft) [reply src/Shared/StackExchangeClientConnection.cs]: Done

**Comment by zachrybaker** [comment]
@stanleysmall-microsoft Can you comment on what is going on with this PR and branch? AFAICT I submitted a PR to restore the ability to serialize to whatever format necessary, providing the means to do safe serialization. I also provided an option to publish the session so that subscribers (for example replacement apps using the Microsoft.SystemWebAdapters package) could interop and share session out of process. This is important for the Blazor use case, as session support in Blazor server is not there yet (and may never be).  It seems you guys repurposed my branch to do maintenance work but axed my features and the whole purpose of the branch. Can you explain?

File: test/RedisSessionStateProviderFunctionalTests/RedisSessionStateProvider.FunctionalTests.csproj (hunk 16 above)
Comment (stanleysmall-microsoft) [reply test/RedisSessionStateProviderFunctionalTests/RedisSessionStateProvider.FunctionalTests.csproj:16]: Added

File: test/OutputCacheProviderFunctionalTests/RedisOutputCacheProvider.FunctionalTests.csproj (hunk 13 above)
Comment (stanleysmall-microsoft) [reply test/OutputCacheProviderFunctionalTests/RedisOutputCacheProvider.FunctionalTests.csproj]: Changed

File: src/Shared/ProviderConfiguration.cs (hunk 11 above)
Comment (stanleysmall-microsoft) [reply src/Shared/ProviderConfiguration.cs]: That's how we did it before: https://github.com/Azure/aspnet-redis-providers/blob/802a2c0721adc02002a7261519f86f551953f369/src/Shared/RedisUtility.cs#L22

File: src/RedisSessionStateProvider/KeyGenerator.cs (hunk 6 above)
Comment (stanleysmall-microsoft) [reply src/RedisSessionStateProvider/KeyGenerator.cs]: I added a suffix `_SessionStateItemCollection_MyCustomSerializer` but leave the value as `_SessionStateItemCollection` if the default serializer is used to provide backward compatibility.

**Comment by frankyjquintero** [comment]
@stanleysmall-microsoft @philon-msft Can you comment on what is going on with this PR and branch?

**Comment by stanleysmall-microsoft** [comment]
Hi @frankyjquintero, I intend on merging and releasing this change soon. It is ready for review (in my opinion) and I am waiting for PR feedback.

**Comment by stanleysmall-microsoft** [comment]
After discussing with our team leadership, we’ve decided against merging this PR. Drivers for this decision include:


- This provider is on a retirement path because ASP.NET Core contains new built-in caching functionality that should be used instead. As the package approaches retirement, the priority is on maintaining stability and minimizing changes.
- As an MIT-licensed open source project, users are welcome to fork this code and modify it to add their own custom serialization as needed. In most cases, that approach will yield a simpler and more efficient solution than going through a generalized ISerializer.
- Users who prefer not to modify provider code can use the previous version 4.0.1 of this package, which does support custom serializers.

 

That said, we do appreciate the work that went into this contribution and the sense of community that it demonstrates. We look forward to working with you to evolve the caching story in the ASP.NET Core space. Recent developments there are simplifying cache integration while dramatically improving performance:

- New caching abstractions: https://devblogs.microsoft.com/dotnet/caching-abstraction-improvements-in-aspnetcore/
- Hybrid cache: https://www.youtube.com/watch?v=TDyZc11cJfA
